Framework/Source/Graphics/Program/ParameterBlock.h

```diff
@@ -0,0 +1,272 @@
+        */
+        UnorderedAccessView::SharedPtr getUav(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex) const;
+
+        /** Bind a sampler to the program in the global namespace.
+        \param[in] name The name of the sampler object in the shader
+        \param[in] pSampler The sampler object to bind
+        \return false if the sampler was not found in the program, otherwise true
+        */
+        bool setSampler(const std::string& name, const Sampler::SharedPtr& pSampler);
+
+        /** Bind a sampler to the program in the global namespace.
+        \param[in] setIndex The set-index in the block
+        \param[in] descIndex The descriptor index inside the set
+        \return false if the sampler was not found in the program, otherwise true
+        */
+        bool setSampler(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex, const Sampler::SharedPtr& pSampler);
+
+        /** Gets a sampler object.
+        \return If the index is valid, a shared pointer to the sampler. Otherwise returns nullptr
+        */
+        Sampler::SharedPtr getSampler(const std::string& name) const;
+
+        /** Gets a sampler object.
+        \param[in] setIndex The set-index in the block
+        \param[in] descIndex The descriptor index inside the set
+        \return If the index is valid, a shared pointer to the sampler. Otherwise returns nullptr
+        */
+        Sampler::SharedPtr getSampler(uint32_t regSpace, uint32_t baseRegIndex, uint32_t arrayIndex) const;
+
+        /** Get the program reflection interface
+        */
+        ParameterBlockReflection::SharedConstPtr getReflection() const { return mpReflector; }
+
+        bool prepareForDraw(CopyContext* pContext, const RootSignature* pRootSig);
+
+        struct RootData
+        {
+            RootData() = default;
+            RootData(uint32_t root, uint32_t range) : rootIndex(root), rangeIndex(range) {}
+            uint32_t rootIndex = uint32_t(-1);
+            uint32_t rangeIndex = uint32_t(-1);
+        };
+
+        template<typename ViewType>
+        struct ResourceData
+        {
+            ResourceData(const RootData& data) : rootData(data) {}
+            typename ViewType::SharedPtr pView = nullptr;
+            Resource::SharedPtr pResource = nullptr;
+            RootData rootData;
+        };
+
+        template<>
+        struct ResourceData<Sampler>
+        {
+            ResourceData(const RootData& data) : rootData(data) {}
+            Sampler::SharedPtr pSampler;
+            RootData rootData;
+        };
+
+
+        struct RootSet
+        {
+            mutable std::shared_ptr<DescriptorSet> pDescSet;
+            mutable bool dirty = false;
+        };
+
+        union BindLocation
+        {
+            BindLocation() = default;
+            BindLocation(uint32_t space, uint32_t index) : baseRegIndex(index), regSpace(space) {}
+            struct
```

Comment:
Does this anonymous `struct` declaration do anything?

Reply:
It is required for the union.